Change observed CCE volume data format.

duetosymmetry opened the pull request:
## Proposed changes

Rather than putting each radial shell into its own subgroup, each complex field is unraveled into a single complex vector. At each observation event, we have a complex vector where compactified radius indexes the slowest, followed by \ell number, followed by azimuthal m number, followed by real/imaginary parts interleaved.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

duetosymmetry committed:
Change observed CCE volume data format to use actual volume format.

The volume format keeps track of the extents of the (unraveled) complex data.
For each ObservationId, each field's dataset is indexed by compactified radial
index (slowest), followed by \ell number, followed by azimuthal m number,
followed by real/imaginary parts interleaved.

The data now go into the file specified by VolumeFileName instead of
ReductionFileName. At the same time I also added an option SubgroupName to
control the group within this file that holds all the volume data. There are up
to three volume subfiles within the subgroup,
  /<SubgroupName>/VolumeData.vol : all normal spin-weighted scalars in volume
  /<SubgroupName>/OneMinusY.vol  : for the compactified radial coordinate, only
    depends on radius and not on angle.
  /<SubgroupName>/InertialRetardedTime.vol : for the complex modal coefficients
    of the inertial retarded time, only depends on angle and not on radius.

This change required making it possible for WriteVolumeData to be invoked as a
local synchronous action.

duetosymmetry commented on src/Evolution/Systems/Cce/Events/ObserveFields.hpp:
```diff
@@ -408,35 +351,66 @@ class ObserveFields : public Event {
+      //      ++i) {
+      //   data_to_write[2 * i    ] = real(goldberg_modes.data()[i]);
+      //   data_to_write[2 * i + 1] = imag(goldberg_modes.data()[i]);
+      // }
```
Oops, I'll take this out in a bit!

duetosymmetry commented:
When trying to actually dump a bunch of volume data (not just tiny tests), I ran into an H5 issue:
```
HDF5-DIAG: Error detected in HDF5 (1.12.3) thread 0:
  #000: H5A.c line 297 in H5Acreate2(): unable to create attribute
    major: Attribute
    minor: Unable to initialize object
  #001: H5VLcallback.c line 988 in H5VL_attr_create(): attribute create failed
    major: Virtual Object Layer
    minor: Unable to create file
  #002: H5VLcallback.c line 955 in H5VL__attr_create(): attribute create failed
    major: Virtual Object Layer
    minor: Unable to create file
  #003: H5VLnative_attr.c line 75 in H5VL__native_attr_create(): unable to create attribute
    major: Attribute
    minor: Unable to initialize object
  #004: H5Aint.c line 267 in H5A__create(): unable to create attribute in object header
    major: Attribute
    minor: Unable to insert object
  #005: H5Oattribute.c line 316 in H5O__attr_create(): unable to create new attribute in header
    major: Attribute
    minor: Unable to insert object
  #006: H5Omessage.c line 204 in H5O__msg_append_real(): unable to create new message
    major: Object header
    minor: No space available for allocation
  #007: H5Omessage.c line 1845 in H5O__msg_alloc(): unable to allocate space for message
    major: Object header
    minor: Unable to initialize object
  #008: H5Oalloc.c line 1290 in H5O__alloc(): object header message is too large
    major: Object header
    minor: Unable to initialize object
```
There is some H5 limitation on the size of "headers" (legends)

duetosymmetry commented:
Macos build is failing because of #6569

wthrowe commented:
> There is some H5 limitation on the size of "headers" (legends)

No particular thoughts on what we should do about it, but, from the source, the limit is something related to
```
#define H5O_MESG_MAX_SIZE 65536 /*max obj header message size        */
```
Is your legend actually approaching 65 KB?

knelli2 commented:
@wthrowe Yes the legend was that large because all radii and all modes were put into the same dat file/legend. Our solution is to just use the volume data writing instead of dat files since that was intended for more data

duetosymmetry commented:
Address #6544

duetosymmetry commented:
For visibility:

The volume format keeps track of the extents of the (unraveled) complex data.
For each ObservationId, each field's dataset is indexed by compactified radial
index (slowest), followed by \ell number, followed by azimuthal m number,
followed by real/imaginary parts interleaved.

The data now go into the file specified by VolumeFileName instead of
ReductionFileName. At the same time I also added an option SubgroupName to
control the group within this file that holds all the volume data. There are up
to three volume subfiles within the subgroup,

    /<SubgroupName>/VolumeData.vol : all normal spin-weighted scalars in volume
    /<SubgroupName>/OneMinusY.vol  : for the compactified radial coordinate, only
      depends on radius and not on angle.
    /<SubgroupName>/InertialRetardedTime.vol : for the complex modal coefficients
    of the inertial retarded time, only depends on angle and not on radius.

This change required making it possible for WriteVolumeData to be invoked as a
local synchronous action.

This change also required updating the meaning of SkipColumns for
CheckOutputFiles.py. Instead of assuming that the dataset we check is
rank-2 (e.g. a matrix), the "columns" now mean the last index, regardless of
rank. Therefore this will make sense for all past invocations, but also for
checking volume data, which has been unraveled into a rank-1 data vector.

duetosymmetry committed:
CheckOutputFiles improvements:

1. Allow to be able to find subfiles using python regexps
2. Instead of assuming that the dataset we check is
rank-2 (e.g. a matrix), the "columns" now mean the last index, regardless of
rank. Therefore this will make sense for all past invocations, but also for
checking volume data, which has been unraveled into a rank-1 data vector.

knelli2 commented on src/Evolution/Systems/Cce/Events/ObserveFields.hpp:
```diff
@@ -99,43 +100,54 @@ std::string name() {
+ * `/CceVolumeData/InertialRetardedTime.vol` for the inertial retarded time; and
+ * it would contain `/CceVolumeData/OneMinusY.vol` for the compactified radial
+ * coordinate. The structure of the .vol subfiles is the same as that for DG
+ * volume data.
```
Add a comment that the extents are different. I.e. for the modal dimensions, the extents are just ell and em.

duetosymmetry replied:
Updated comments and pushed.

knelli2 reviewed (changes requested):
Just a small dox thing. You can squash directly.

duetosymmetry commented:
Are you aware of any other docs that should be updated? Should the CCE tutorial say anything about volume data, or no?

knelli2 commented:
The only other place would be the CCE tutorial. Yeah I think it'd be worth adding a small bullet in the "General Options" subsection under "Input file for CCE" in the CCE tutorial page. Basically just saying that you can dump certain volume quantities with the `ObserveFields` event such as J or Psi0. No need to go too in depth

duetosymmetry commented:
Added a bullet point to the tutorial, referring to where the main documentation is for Cce::Events::ObserveFields. I also edited the yaml file that the tutorial points to as an example, with a commented ObserveFields group.

duetosymmetry commented:
Didn't mean to dismiss the review... just rebased on latest develop to get rid of the failing test
